internal/cephfs/controllerserver.go

```diff
@@ -294,3 +444,309 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
 		NodeExpansionRequired: false,
 	}, nil
 }
+
+// CreateSnapshot creates the snapshot in backend and stores metadata
+// in store
+// nolint:gocyclo // golangci-lint did not catch this earlier, needs to get fixed late
+func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	if err := cs.validateSnapshotReq(ctx, req); err != nil {
+		return nil, err
+	}
+	cr, err := util.NewAdminCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, err
+	}
+	defer cr.DeleteCredentials()
+
+	clusterData, err := getClusterInformation(req.GetParameters())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	requestName := req.GetName()
+	sourceVolID := req.GetSourceVolumeId()
+	// Existence and conflict checks
+	if acquired := cs.SnapshotLocks.TryAcquire(requestName); !acquired {
+		klog.Errorf(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), requestName)
+		return nil, status.Errorf(codes.Aborted, util.SnapshotOperationAlreadyExistsFmt, requestName)
+	}
+	defer cs.SnapshotLocks.Release(requestName)
+
+	if err = cs.OperationLocks.GetSnapshotCreateLock(sourceVolID); err != nil {
+		klog.Error(util.Log(ctx, err.Error()))
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+
+	defer cs.OperationLocks.ReleaseSnapshotCreateLock(sourceVolID)
+
+	// Find the volume using the provided VolumeID
+	parentVolOptions, vid, err := newVolumeOptionsFromVolID(ctx, sourceVolID, nil, req.GetSecrets())
+	if err != nil {
+		if errors.Is(err, util.ErrPoolNotFound) {
+			klog.Warningf(util.Log(ctx, "failed to get backend volume for %s: %v"), sourceVolID, err)
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+
+		if errors.Is(err, ErrVolumeNotFound) {
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if clusterData.ClusterID != parentVolOptions.ClusterID {
+		return nil, status.Errorf(codes.InvalidArgument, "requested cluster id %s not matching subvolume cluster id %s", clusterData.ClusterID, parentVolOptions.ClusterID)
+	}
+
+	cephfsSnap := genSnapFromOptions(ctx, req)
+	if cephfsSnap == nil {
+		return nil, status.Error(codes.Internal, err.Error())
```

Comment:
This will panic which err.Error we are returning here?

Reply:
done..thanks